config.org
#+TITLE: James' Doom Emacs Config
#+AUTHOR: James Stone
#+DESCRIPTION: James' personal Doom Emacs config.
#+STARTUP: showeverything

* TABLE OF CONTENTS :toc:
- [[#about-this-config][ABOUT THIS CONFIG]]
- [[#bookmarks-and-buffers][BOOKMARKS AND BUFFERS]]
- [[#centaur-tabs][CENTAUR-TABS]]
- [[#dired][DIRED]]
- [[#doom-theme][DOOM THEME]]
- [[#elfeed][ELFEED]]
- [[#evaluate-elisp-expressions][EVALUATE ELISP EXPRESSIONS]]
- [[#eww][EWW]]
- [[#fonts][FONTS]]
- [[#line-settings][LINE SETTINGS]]
- [[#mouse-support][MOUSE SUPPORT]]
- [[#mu4e][MU4E]]
  - [[#setting-up-environment-with-main-account][Setting Up Environment With Main Account]]
  - [[#setting-up-multiple-accounts][Setting Up Multiple Accounts]]
  - [[#function-to-facilitate-switching-between-accounts][Function To Facilitate Switching Between Accounts]]
- [[#open-specific-files][OPEN SPECIFIC FILES]]
- [[#org-mode][ORG MODE]]
  - [[#roam][Roam]]
- [[#projectile][PROJECTILE]]
- [[#registers][REGISTERS]]
- [[#shells][SHELLS]]
- [[#slack][SLACK]]
- [[#splits][SPLITS]]
- [[#telega][TELEGA]]
- [[#winner-mode][WINNER MODE]]

* ABOUT THIS CONFIG 
This is my personal Doom Emacs config. It has been modified from [[https://gitlab.com/dwt1/dotfiles/-/blob/master/.doom.d/config.org][DT's Doom Emacs config]]. Doom Emacs is a distribution of Emacs that uses the "evil" keybindings (Vim keybindings) and includes a number of nice extensions and a bit of configuration out of the box.

* BOOKMARKS AND BUFFERS
Doom Emacs uses 'SPC b' for keybindings related to bookmarks and buffers.  Bookmarks are somewhat like registers in that they record positions you can jump to.  Unlike registers, they have long names, and they persist automatically from one Emacs session to the next. The prototypical use of bookmarks is to record where you were reading in various files.  Regarding /buffers/, the text you are editing in Emacs resides in an object called a /buffer/. Each time you visit a file, a buffer is used to hold the file’s text. Each time you invoke Dired, a buffer is used to hold the directory listing.

#+BEGIN_SRC emacs-lisp
(map! :leader
      :desc "List bookmarks"
      "b L" #'list-bookmarks
      :leader
      :desc "Save current bookmarks to bookmark file"
      "b w" #'bookmark-save)
#+END_SRC

* CENTAUR-TABS
To use tabs in Doom Emacs, be sure to uncomment "tabs" in Doom's init.el.  Displays tabs at the top of the window similar to tabbed web browsers such as Firefox.  I don't actually use tabs in Emacs.  I placed this in my config to help others who may want tabs.  In the default configuration of Doom Emacs, 'SPC t' is used for "toggle" keybindings, so I choose 'SPC t c' to toggle centaur-tabs.  The "g" prefix for keybindings is used for a bunch of evil keybindings in Doom, but "g" plus the arrow keys were not used, so I thought I would bind those for tab navigation.  But I did leave the default "g t" and "g T" intact if you prefer to use those for centaur-tabs-forward/backward.

| COMMAND                     | DESCRIPTION               | KEYBINDING       |
|-----------------------------+---------------------------+------------------|
| centaur-tabs-mode           | /Toggle tabs globally/      | SPC t c          |
| centaur-tabs-local-mode     | /Toggle tabs local display/ | SPC t C          |
| centaur-tabs-forward        | /Next tab/                  | g <right> or g t |
| centaur-tabs-backward       | /Previous tab/              | g <left> or g T  |
| centaur-tabs-forward-group  | /Next tab group/            | g <down>         |
| centaur-tabs-backward-group | /Previous tab group/        | g <up>           |

#+BEGIN_SRC emacs-lisp
(setq centaur-tabs-set-bar 'over
      centaur-tabs-set-icons t
      centaur-tabs-gray-out-icons 'buffer
      centaur-tabs-height 24
      centaur-tabs-set-modified-marker t
      centaur-tabs-style "bar"
      centaur-tabs-modified-marker "•")
(map! :leader
      :desc "Toggle tabs globally"
      "t c" #'centaur-tabs-mode
      :leader
      :desc "Toggle tabs local display"
      "t C" #'centaur-tabs-local-mode)
(evil-define-key 'normal centaur-tabs-mode-map (kbd "g <right>") 'centaur-tabs-forward        ; default Doom binding is 'g t'
                                               (kbd "g <left>")  'centaur-tabs-backward       ; default Doom binding is 'g T'
                                               (kbd "g <down>")  'centaur-tabs-forward-group
                                               (kbd "g <up>")    'centaur-tabs-backward-group)
#+END_SRC

* DIRED
Dired is the file manager within Emacs.  Below, I setup keybindings for image previews (peep-dired).  Doom Emacs does not use 'SPC d' for any of its keybindings, so I've chosen the format of 'SPC d' plus 'key'.

| COMMAND                                   | DESCRIPTION                                     | KEYBINDING |
|-------------------------------------------+-------------------------------------------------+------------|
| dired                                     | /Open dired file manager/                         | SPC d d    |
| dired-jump                                | /Jump to current directory in dired/              | SPC d j    |
| (in dired) peep-dired                     | /Toggle image previews within dired/              | SPC d p    |
| (in dired) dired-view-file                | /View file in dired/                              | SPC d v    |
| (in dired) dired-up-directory             | /Go up in the directory tree/                     | h          |
| (in dired) dired-find-file                | /Go down in the directory tree (or open if file)/ | l          |
| (in peep-dired-mode) peep-dired-next-file | /Move to next file in peep-dired-mode/            | j          |
| (in peep-dired-mode) peep-dired-prev-file | /Move to previous file in peep-dired-mode/        | k          |

#+BEGIN_SRC emacs-lisp
(map! :leader
      :desc "Dired"
      "d d" #'dired
      :leader
      :desc "Dired jump to current"
      "d j" #'dired-jump
      (:after dired
        (:map dired-mode-map
         :leader
         :desc "Peep-dired image previews"
         "d p" #'peep-dired
         :leader
         :desc "Dired view file"
         "d v" #'dired-view-file)))
;; Make 'h' and 'l' go back and forward in dired. Much faster to navigate the directory structure!
(evil-define-key 'normal dired-mode-map
  (kbd "h") 'dired-up-directory
  (kbd "l") 'dired-open-file) ; use dired-find-file instead if not using dired-open package
;; If peep-dired is enabled, you will get image previews as you go up/down with 'j' and 'k'
(evil-define-key 'normal peep-dired-mode-map
  (kbd "j") 'peep-dired-next-file
  (kbd "k") 'peep-dired-prev-file)
(add-hook 'peep-dired-hook 'evil-normalize-keymaps)
;; Get file icons in dired
(add-hook 'dired-mode-hook 'all-the-icons-dired-mode)
;; With dired-open plugin, you can launch external programs for certain extensions
;; For example, I set all .png files to open in 'sxiv' and all .mp4 files to open in 'mpv'
(setq dired-open-extensions '(("gif" . "sxiv")
                              ("jpg" . "sxiv")
                              ("png" . "sxiv")
                              ("mkv" . "mpv")
                              ("mp4" . "mpv")))
#+END_SRC

* DOOM THEME
Setting the theme to doom-one.  To try out new themes, I set a keybinding for counsel-load-theme with 'SPC h t'.

#+BEGIN_SRC emacs-lisp
(setq doom-theme 'doom-solarized-light)
(map! :leader
      :desc "Load new theme"
      "h t" #'counsel-load-theme)
#+END_SRC

* ELFEED
An RSS newsfeed reader for Emacs.

#+BEGIN_SRC emacs-lisp
(require 'elfeed-goodies)
(elfeed-goodies/setup)
(setq elfeed-goodies/entry-pane-size 0.5)

(defun elfeed-goodies/entry-line-draw (entry)
  "Print ENTRY to the buffer."

  (let* ((title (or (elfeed-meta entry :title) (elfeed-entry-title entry) ""))
  (date (elfeed-search-format-date (elfeed-entry-date entry)))
  (title-faces (elfeed-search--faces (elfeed-entry-tags entry)))
  (feed (elfeed-entry-feed entry))
  (feed-title
   (when feed
     (or (elfeed-meta feed :title) (elfeed-feed-title feed))))
  (tags (mapcar #'symbol-name (elfeed-entry-tags entry)))
  (tags-str (concat "[" (mapconcat 'identity tags ",") "]"))
  (title-width (- (window-width) elfeed-goodies/feed-source-column-width
    elfeed-goodies/tag-column-width 4))
  (title-column (elfeed-format-column
   title (elfeed-clamp
          elfeed-search-title-min-width
          title-width
          title-width)
   :left))
  (tag-column (elfeed-format-column
        tags-str (elfeed-clamp (length tags-str)
          elfeed-goodies/tag-column-width
          elfeed-goodies/tag-column-width)
        :left))
  (feed-column (elfeed-format-column
         feed-title (elfeed-clamp elfeed-goodies/feed-source-column-width
      elfeed-goodies/feed-source-column-width
      elfeed-goodies/feed-source-column-width)
         :left)))

    (if (>= (window-width) (* (frame-width) elfeed-goodies/wide-threshold))
 (progn
   (insert (propertize title 'face title-faces 'kbd-help title) " ")
   (insert (propertize title 'face title-faces 'kbd-help title) " ")
   (insert (propertize date 'face 'elfeed-search-date-face) " ")
   (insert (propertize feed-column 'face 'elfeed-search-feed-face) " ")
   (insert (propertize tag-column 'face 'elfeed-search-tag-face) " ")
    ))
    )
)
;; (defun ap/elfeed-goodies/entry-line-draw (entry)
;;   "My version of this function.  Prints ENTRY to the buffer."

;;   (cl-flet ((add-faces (str &rest faces)
;;                        (dolist (face faces str)
;;                          (add-face-text-property 0 (length str)
;;                                                  face 'append str)))
;;             (tags (entry)
;;                   (seq-difference (--map (substring-no-properties (symbol-name it))
;;                                          (elfeed-entry-tags entry))
;;                                   '("unread" "starred"))))
;;     (let* ( ;; Choose color and faces first
;;            ;; See https://www.reddit.com/r/emacs/comments/7a976a/face_applied_to_result_of_symbolname_becomes/
;;            (site (pocket-reader--url-domain (elfeed-entry-link entry)))
;;            (hash (rainbow-identifiers--hash-function site))
;;            (site-face (rainbow-identifiers-cie-l*a*b*-choose-face hash))
;;            (title-faces (elfeed-search--faces (elfeed-entry-tags entry)))

;;            ;; Feed
;;            (feed (elfeed-entry-feed entry))
;;            (feed-title (when feed
;;                          (or (elfeed-meta feed :title) (elfeed-feed-title feed))))
;;            (feed-width elfeed-goodies/feed-source-column-width)
;;            (feed-column (elfeed-format-column feed-title feed-width :left))
;;            (feed-column (apply #'add-faces feed-column site-face title-faces))

;;            ;; Tags before title (so title can use the width of the tags column for this item)
;;            (tags (tags entry))
;;            (tags-str (s-join "," tags))
;;            ;; Use raw tag list to check for starred
;;            (starred-p (member 'starred (elfeed-entry-tags entry)))
;;            (tags-width (min (length tags-str)
;;                             elfeed-goodies/tag-column-width))
;;            (tag-column (elfeed-format-column tags-str tags-width :right))
;;            (tag-column (apply #'add-faces tag-column site-face title-faces))

;;            ;; Title
;;            (title (or (elfeed-meta entry :title) (elfeed-entry-title entry) ""))
;;            (title-width (- (window-width) feed-width tags-width 4))
;;            (title-column (elfeed-format-column (truncate-string-to-width title title-width nil nil 'ellipsis) title-width :left))
;;            (title-column (apply #'add-faces title-column site-face title-faces)))

;;       (insert feed-column " ")
;;       (insert (if starred-p
;;                   (propertize "*"
;;                               'face 'pocket-reader-favorite-star)
;;                 " ")
;;               " ")
;;       (insert (propertize title-column 'kbd-help title) " ")
;;       (insert tag-column))))

;;      (advice-add #'elfeed-goodies/entry-line-draw :override #'ap/elfeed-goodies/entry-line-draw)

(add-hook 'elfeed-search-mode-hook 'elfeed-update)
(add-hook 'elfeed-new-entry-hook
          (elfeed-make-tagger :before "2 weeks ago"
                              :remove 'unread))
(add-hook 'elfeed-show-mode-hook 'visual-line-mode)
(evil-define-key 'normal elfeed-show-mode-map
  (kbd "J") 'elfeed-goodies/split-show-next
  (kbd "K") 'elfeed-goodies/split-show-prev)
(evil-define-key 'normal elfeed-search-mode-map
  (kbd "J") 'elfeed-goodies/split-show-next
  (kbd "K") 'elfeed-goodies/split-show-prev)
(setq elfeed-feeds (quote
                    (("https://www.reddit.com/r/linux.rss" reddit linux)
                     ("https://www.reddit.com/r/commandline.rss" reddit linux)
                     ("https://stallman.org/rss/rss.xml" news foss gnu)
                     ("https://www.reddit.com/r/emacs.rss" reddit linux)
                     ("https://www.gamingonlinux.com/article_rss.php" gaming linux)
                     ("https://hackaday.com/blog/feed/" hackaday linux)
                     ("https://opensource.com/feed" opensource linux)

config.el
(map! :leader
      :desc "List bookmarks"
      "b L" #'list-bookmarks
      :leader
      :desc "Save current bookmarks to bookmark file"
      "b w" #'bookmark-save)

(setq centaur-tabs-set-bar 'over
      centaur-tabs-set-icons t
      centaur-tabs-gray-out-icons 'buffer
      centaur-tabs-height 24
      centaur-tabs-set-modified-marker t
      centaur-tabs-style "bar"
      centaur-tabs-modified-marker "•")
(map! :leader
      :desc "Toggle tabs globally"
      "t c" #'centaur-tabs-mode
      :leader
      :desc "Toggle tabs local display"
      "t C" #'centaur-tabs-local-mode)
(evil-define-key 'normal centaur-tabs-mode-map (kbd "g <right>") 'centaur-tabs-forward        ; default Doom binding is 'g t'
                                               (kbd "g <left>")  'centaur-tabs-backward       ; default Doom binding is 'g T'
                                               (kbd "g <down>")  'centaur-tabs-forward-group
                                               (kbd "g <up>")    'centaur-tabs-backward-group)

(map! :leader
      :desc "Dired"
      "d d" #'dired
      :leader
      :desc "Dired jump to current"
      "d j" #'dired-jump
      (:after dired
        (:map dired-mode-map
         :leader
         :desc "Peep-dired image previews"
         "d p" #'peep-dired
         :leader
         :desc "Dired view file"
         "d v" #'dired-view-file)))
;; Make 'h' and 'l' go back and forward in dired. Much faster to navigate the directory structure!
(evil-define-key 'normal dired-mode-map
  (kbd "h") 'dired-up-directory
  (kbd "l") 'dired-open-file) ; use dired-find-file instead if not using dired-open package
;; If peep-dired is enabled, you will get image previews as you go up/down with 'j' and 'k'
(evil-define-key 'normal peep-dired-mode-map
  (kbd "j") 'peep-dired-next-file
  (kbd "k") 'peep-dired-prev-file)
(add-hook 'peep-dired-hook 'evil-normalize-keymaps)
;; Get file icons in dired
(add-hook 'dired-mode-hook 'all-the-icons-dired-mode)
;; With dired-open plugin, you can launch external programs for certain extensions
;; For example, I set all .png files to open in 'sxiv' and all .mp4 files to open in 'mpv'
(setq dired-open-extensions '(("gif" . "sxiv")
                              ("jpg" . "sxiv")
                              ("png" . "sxiv")
                              ("mkv" . "mpv")
                              ("mp4" . "mpv")))

(setq doom-theme 'doom-solarized-light)
(map! :leader
      :desc "Load new theme"
      "h t" #'counsel-load-theme)

(require 'elfeed-goodies)
(elfeed-goodies/setup)
(setq elfeed-goodies/entry-pane-size 0.5)

(defun elfeed-goodies/entry-line-draw (entry)
  "Print ENTRY to the buffer."

  (let* ((title (or (elfeed-meta entry :title) (elfeed-entry-title entry) ""))
  (date (elfeed-search-format-date (elfeed-entry-date entry)))
  (title-faces (elfeed-search--faces (elfeed-entry-tags entry)))
  (feed (elfeed-entry-feed entry))
  (feed-title
   (when feed
     (or (elfeed-meta feed :title) (elfeed-feed-title feed))))
  (tags (mapcar #'symbol-name (elfeed-entry-tags entry)))
  (tags-str (concat "[" (mapconcat 'identity tags ",") "]"))
  (title-width (- (window-width) elfeed-goodies/feed-source-column-width
    elfeed-goodies/tag-column-width 4))
  (title-column (elfeed-format-column
   title (elfeed-clamp
          elfeed-search-title-min-width
          title-width
          title-width)
   :left))
  (tag-column (elfeed-format-column
        tags-str (elfeed-clamp (length tags-str)
          elfeed-goodies/tag-column-width
          elfeed-goodies/tag-column-width)
        :left))
  (feed-column (elfeed-format-column
         feed-title (elfeed-clamp elfeed-goodies/feed-source-column-width
      elfeed-goodies/feed-source-column-width
      elfeed-goodies/feed-source-column-width)
         :left)))

    (if (>= (window-width) (* (frame-width) elfeed-goodies/wide-threshold))
 (progn
   (insert (propertize title 'face title-faces 'kbd-help title) " ")
   (insert (propertize title 'face title-faces 'kbd-help title) " ")
   (insert (propertize date 'face 'elfeed-search-date-face) " ")
   (insert (propertize feed-column 'face 'elfeed-search-feed-face) " ")
   (insert (propertize tag-column 'face 'elfeed-search-tag-face) " ")
    ))
    )
)
;; (defun ap/elfeed-goodies/entry-line-draw (entry)
;;   "My version of this function.  Prints ENTRY to the buffer."

;;   (cl-flet ((add-faces (str &rest faces)
;;                        (dolist (face faces str)
;;                          (add-face-text-property 0 (length str)
;;                                                  face 'append str)))
;;             (tags (entry)
;;                   (seq-difference (--map (substring-no-properties (symbol-name it))
;;                                          (elfeed-entry-tags entry))
;;                                   '("unread" "starred"))))
;;     (let* ( ;; Choose color and faces first
;;            ;; See https://www.reddit.com/r/emacs/comments/7a976a/face_applied_to_result_of_symbolname_becomes/
;;            (site (pocket-reader--url-domain (elfeed-entry-link entry)))
;;            (hash (rainbow-identifiers--hash-function site))
;;            (site-face (rainbow-identifiers-cie-l*a*b*-choose-face hash))
;;            (title-faces (elfeed-search--faces (elfeed-entry-tags entry)))

;;            ;; Feed
;;            (feed (elfeed-entry-feed entry))
;;            (feed-title (when feed
;;                          (or (elfeed-meta feed :title) (elfeed-feed-title feed))))
;;            (feed-width elfeed-goodies/feed-source-column-width)
;;            (feed-column (elfeed-format-column feed-title feed-width :left))
;;            (feed-column (apply #'add-faces feed-column site-face title-faces))

;;            ;; Tags before title (so title can use the width of the tags column for this item)
;;            (tags (tags entry))
;;            (tags-str (s-join "," tags))
;;            ;; Use raw tag list to check for starred
;;            (starred-p (member 'starred (elfeed-entry-tags entry)))
;;            (tags-width (min (length tags-str)
;;                             elfeed-goodies/tag-column-width))
;;            (tag-column (elfeed-format-column tags-str tags-width :right))
;;            (tag-column (apply #'add-faces tag-column site-face title-faces))

;;            ;; Title
;;            (title (or (elfeed-meta entry :title) (elfeed-entry-title entry) ""))
;;            (title-width (- (window-width) feed-width tags-width 4))
;;            (title-column (elfeed-format-column (truncate-string-to-width title title-width nil nil 'ellipsis) title-width :left))
;;            (title-column (apply #'add-faces title-column site-face title-faces)))

;;       (insert feed-column " ")
;;       (insert (if starred-p
;;                   (propertize "*"
;;                               'face 'pocket-reader-favorite-star)
;;                 " ")
;;               " ")
;;       (insert (propertize title-column 'kbd-help title) " ")
;;       (insert tag-column))))

;;      (advice-add #'elfeed-goodies/entry-line-draw :override #'ap/elfeed-goodies/entry-line-draw)

(add-hook 'elfeed-search-mode-hook 'elfeed-update)
(add-hook 'elfeed-new-entry-hook
          (elfeed-make-tagger :before "2 weeks ago"
                              :remove 'unread))
(add-hook 'elfeed-show-mode-hook 'visual-line-mode)
(evil-define-key 'normal elfeed-show-mode-map
  (kbd "J") 'elfeed-goodies/split-show-next
  (kbd "K") 'elfeed-goodies/split-show-prev)
(evil-define-key 'normal elfeed-search-mode-map
  (kbd "J") 'elfeed-goodies/split-show-next
  (kbd "K") 'elfeed-goodies/split-show-prev)
(setq elfeed-feeds (quote
                    (("https://www.reddit.com/r/linux.rss" reddit linux)
                     ("https://www.reddit.com/r/commandline.rss" reddit linux)
                     ("https://stallman.org/rss/rss.xml" news foss gnu)
                     ("https://www.reddit.com/r/emacs.rss" reddit linux)
                     ("https://www.gamingonlinux.com/article_rss.php" gaming linux)
                     ("https://hackaday.com/blog/feed/" hackaday linux)
                     ("https://opensource.com/feed" opensource linux)
                     ("https://linux.softpedia.com/backend.xml" softpedia linux)
                     ("https://itsfoss.com/feed/" itsfoss linux)
                     ("https://www.zdnet.com/topic/linux/rss.xml" zdnet linux)
                     ("https://www.phoronix.com/rss.php" phoronix linux)
                     ("http://feeds.feedburner.com/d0od" omgubuntu linux)
                     ("https://www.computerworld.com/index.rss" computerworld linux)
                     ("https://www.networkworld.com/category/linux/index.rss" networkworld linux)
                     ("https://www.techrepublic.com/rssfeeds/topic/open-source/" techrepublic linux)
                     ("https://betanews.com/feed" betanews linux)
                     ("http://lxer.com/module/newswire/headlines.rss" lxer linux)
                     ("https://distrowatch.com/news/dwd.xml" distrowatch linux))))

(map! :leader
      :desc "Evaluate elisp in buffer"
      "e b" #'eval-buffer
      :leader
      :desc "Evaluate defun"
      "e d" #'eval-defun
      :leader
      :desc "Evaluate elisp expression"
      "e e" #'eval-expression
      :leader
      :desc "Evaluate last sexpression"
      "e l" #'eval-last-sexp
      :leader
      :desc "Evaluate elisp in region"
      "e r" #'eval-region)

(setq browse-url-browser-function 'eww-browse-url)
(map! :leader
      :desc "Eww web browser"
      "e w" #'eww
      :leader
      :desc "Eww reload page"
      "e R" #'eww-reload
      :leader
      :desc "Search web for text between BEG/END"
      "s w" #'eww-search-words)

(setq doom-font (font-spec :family "Mononoki Nerd Font" :size 15)
      doom-variable-pitch-font (font-spec :family "Mononoki Nerd Font" :size 15)
      doom-big-font (font-spec :family "Mononoki Nerd Font" :size 24))
(after! doom-themes
  (setq doom-themes-enable-bold t
        doom-themes-enable-italic t))
(custom-set-faces!
  '(font-lock-comment-face :slant italic)
  '(font-lock-keyword-face :slant italic))

(setq display-line-numbers-type t)
(map! :leader
      :desc "Comment or uncomment lines"
      "TAB TAB" #'comment-line
      :leader
      :desc "Toggle line numbers"
      "t l" #'doom/toggle-line-numbers
      :leader
      :desc "Toggle line highlight in frame"
      "t h" #'hl-line-mode
      :leader
      :desc "Toggle line highlight globally"
      "t H" #'global-hl-line-mode
      :leader
      :desc "Toggle truncate lines"
      "t t" #'toggle-truncate-lines)

(xterm-mouse-mode 1)
(global-set-key [mouse-4] (lambda ()
                              (interactive)
                              (scroll-down 1)))
(global-set-key [mouse-5] (lambda ()
                              (interactive)
                              (scroll-up 1)))

(add-to-list 'load-path "/usr/local/share/emacs/site-lisp/mu4e")
;;(require 'smtpmail)
(auth-source-pass-enable)
(setq user-mail-address "jamesmstone@hotmail.com"
      user-full-name  "James Stone"
      ;; I have my mbsyncrc in a different folder on my system, to keep it separate from the
      ;; mbsyncrc available publicly in my dotfiles. You MUST edit the following line.

      auth-source-debug t
      auth-source-do-cache nil
      auth-sources '(password-store)
      mu4e-get-mail-command "mbsync -a"
      mu4e-update-interval  300
      mu4e-main-buffer-hide-personal-addresses t
      message-send-mail-function 'smtpmail-send-it
      starttls-use-gnutls t
      smtpmail-starttls-credentials '(("smtp-mail.outlook.com" 587 nil nil))
      mu4e-sent-folder "/outlook/Sent"
      mu4e-drafts-folder "/outlook/Drafts"
      mu4e-trash-folder "/outlook/Deleted"
      mu4e-maildir-shortcuts
      '(("/outlook/inbox"       . ?i)
        ("/outlook/Sent"        . ?s)
        ("/outlook/Drafts"      . ?d)
        ("/outlook/Deleted"     . ?t)))

(defvar my-mu4e-account-alist
  '(("outlook"
     (mu4e-sent-folder "/outlook/Sent")
     (mu4e-drafts-folder "/outlook/Drafts")
     (mu4e-trash-folder "/outlook/Deleted")
     (mu4e-compose-signature
       (concat
         "Cheers\n"
         "James"))
     (user-mail-address "jamesmstone@hotmail.com")
     (smtpmail-default-smtp-server "smtp-mail.outlook.com")
     (smtpmail-smtp-server "smtp-mail.outlook.com")
     (smtpmail-smtp-user "jamesmstone@hotmail.com")
     (smtpmail-stream-type starttls)
     (smtpmail-smtp-service 587))
    ;; ("exchange"
    ;;  (mu4e-sent-folder "/exchange/Sent Items")
    ;;  (mu4e-drafts-folder "/exchange/Drafts")
    ;;  (mu4e-trash-folder "/exchange/Deleted Items")
    ;;  (mu4e-compose-signature
    ;;    (concat
    ;;      "James Stone\n"
    ;;      "james@renewabeenergyhub.com.au"))
    ;;  (user-mail-address "james.stone@traditionasia.com")
    ;;  (smtpmail-default-smtp-server "smtp.domain.com")
    ;;  (smtpmail-smtp-server "smtp.domain.com")
    ;;  (smtpmail-smtp-user "acc1@domain.com")
    ;;  (smtpmail-stream-type starttls)
    ;;  (smtpmail-smtp-service 587))
    ("gmail"
     (mu4e-sent-folder "/gmail/sent")
     (mu4e-drafts-folder "/gmail/Drafts")
     (mu4e-trash-folder "/gmail/trash")
     (mu4e-compose-signature
       (concat
         "Cheers\n"
         "James\n"))
     (user-mail-address "jamesmstone711@gmail.com")
     (smtpmail-default-smtp-server "smtp-mail.outlook.com")
     (smtpmail-smtp-server "smtp-mail.outlook.com")
     (smtpmail-smtp-user "jamesmstone711@gmail.com")
     (smtpmail-stream-type starttls)
     (smtpmail-smtp-service 587))))

(defun my-mu4e-set-account ()
  "Set the account for composing a message."
  (let* ((account
          (if mu4e-compose-parent-message
              (let ((maildir (mu4e-message-field mu4e-compose-parent-message :maildir)))
                (string-match "/\\(.*?\\)/" maildir)
                (match-string 1 maildir))
            (completing-read (format "Compose with account: (%s) "
                                     (mapconcat #'(lambda (var) (car var))
                                                my-mu4e-account-alist "/"))
                             (mapcar #'(lambda (var) (car var)) my-mu4e-account-alist)
                             nil t nil nil (caar my-mu4e-account-alist))))
         (account-vars (cdr (assoc account my-mu4e-account-alist))))
    (if account-vars
        (mapc #'(lambda (var)
                  (set (car var) (cadr var)))
              account-vars)
      (error "No email account found"))))

(add-hook 'mu4e-compose-pre-hook 'my-mu4e-set-account)

(map! :leader
      :desc "Edit agenda file"
      "- a" #'(lambda () (interactive) (find-file "~/Org/agenda.org"))
      :leader
      :desc "Edit doom config.org"
      "- c" #'(lambda () (interactive) (find-file "~/.doom.d/config.org"))
      :leader
      :desc "Edit eshell aliases"
      "- e" #'(lambda () (interactive) (find-file "~/.doom.d/aliases"))
      :leader
      :desc "Edit doom init.el"
      "- i" #'(lambda () (interactive) (find-file "~/.doom.d/init.el"))
      :leader
      :desc "Edit doom packages.el"
      "- p" #'(lambda () (interactive) (find-file "~/.doom.d/packages.el")))

(after! org
  (add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))
  (setq org-directory "~/Org/"
        org-agenda-files '("~/Org/agenda.org")
        org-default-notes-file (expand-file-name "notes.org" org-directory)
        org-ellipsis " ▼ "
        org-log-done 'time
        org-journal-dir "~/Org/journal/"
        org-journal-date-prefix "#+TITLE: "
        org-journal-date-format "%B %d, %Y (%A) "
        org-journal-time-prefix "* "
        org-journal-file-format "%Y-%m-%d.org"
        org-hide-emphasis-markers t
        ;; ex. of org-link-abbrev-alist in action
        ;; [[arch-wiki:Name_of_Page][Description]]
        org-link-abbrev-alist    ; This overwrites the default Doom org-link-abbrev-list
          '(("google" . "http://www.google.com/search?q=")
            ("arch-wiki" . "https://wiki.archlinux.org/index.php/")
            ("ddg" . "https://duckduckgo.com/?q=")
            ("wiki" . "https://en.wikipedia.org/wiki/"))
        org-todo-keywords        ; This overwrites the default Doom org-todo-keywords
          '((sequence
             "TODO(t)"           ; A task that is ready to be tackled
             "PROJ(p)"           ; A project that contains other tasks
             "WAIT(w)"           ; Something is holding up this task
             "|"                 ; The pipe necessary to separate "active" states and "inactive" states
             "DONE(d)"           ; Task has been completed
             "CANCELLED(c)" )))) ; Task has been cancelled

(defun dt/org-babel-tangle-async (file)
  "Invoke `org-babel-tangle-file' asynchronously."
  (message "Tangling %s..." (buffer-file-name))
  (async-start
   (let ((args (list file)))
  `(lambda ()
        (require 'org)
        ;;(load "~/.emacs.d/init.el")
        (let ((start-time (current-time)))
          (apply #'org-babel-tangle-file ',args)
          (format "%.2f" (float-time (time-since start-time))))))
   (let ((message-string (format "Tangling %S completed after " file)))
     `(lambda (tangle-time)
        (message (concat ,message-string
                         (format "%s seconds" tangle-time)))))))

(defun dt/org-babel-tangle-current-buffer-async ()
  "Tangle current buffer asynchronously."
  (dt/org-babel-tangle-async (buffer-file-name)))

(setq org-roam-directory "~/Org/")

(setq projectile-project-search-path '("~/"))

(map! :leader
      :desc "Copy to register"
      "r c" #'copy-to-register
      :leader
      :desc "Frameset to register"
      "r f" #'frameset-to-register
      :leader
      :desc "Insert contents of register"
      "r i" #'insert-register
      :leader
      :desc "Jump to register"
      "r j" #'jump-to-register
      :leader
      :desc "List registers"
      "r l" #'list-registers
      :leader
      :desc "Number to register"
      "r n" #'number-to-register
      :leader
      :desc "Interactively choose a register"
      "r r" #'counsel-register
      :leader
      :desc "View a register"
      "r v" #'view-register
      :leader
      :desc "Window configuration to register"
      "r w" #'window-configuration-to-register
      :leader
      :desc "Increment register"
      "r +" #'increment-register
      :leader
      :desc "Point to register"
      "r SPC" #'point-to-register)

(setq shell-file-name "/bin/bash"
      eshell-aliases-file "~/.doom.d/aliases"
      eshell-history-size 5000
      eshell-buffer-maximum-lines 5000
      eshell-hist-ignoredups t
      eshell-scroll-to-bottom-on-input t
      eshell-destroy-buffer-when-process-dies t
      eshell-visual-commands'("bash" "htop" "ssh" "top" "zsh")
      vterm-max-scrollback 5000)
(map! :leader
      :desc "Counsel eshell history"
      "e h" #'counsel-esh-history)

(use-package slack
  :commands (slack-start)
  :init
  (setq slack-buffer-emojify t) ;; if you want to enable emoji, default nil
  (setq slack-prefer-current-team t)
  :config
  (slack-register-team
   :name "tradsasiasoftware"
   :default t
   :token (+pass-get-field "other/Slack_tradition_asia" "token" )
   :cookie (+pass-get-field "other/Slack_tradition_asia" "cookie")
   ;; :subscribed-channels '(test-rename rrrrr)
   :full-and-display-names t)

  (evil-define-key 'normal slack-info-mode-map
    ",u" 'slack-room-update-messages)
  (evil-define-key 'normal slack-mode-map
    ",c" 'slack-buffer-kill
    ",ra" 'slack-message-add-reaction
    ",rr" 'slack-message-remove-reaction
    ",rs" 'slack-message-show-reaction-users
    ",pl" 'slack-room-pins-list
    ",pa" 'slack-message-pins-add
    ",pr" 'slack-message-pins-remove
    ",mm" 'slack-message-write-another-buffer
    ",me" 'slack-message-edit
    ",md" 'slack-message-delete
    ",u" 'slack-room-update-messages
    ",2" 'slack-message-embed-mention
    ",3" 'slack-message-embed-channel
    "\C-n" 'slack-buffer-goto-next-message
    "\C-p" 'slack-buffer-goto-prev-message)
   (evil-define-key 'normal slack-edit-message-mode-map
    ",k" 'slack-message-cancel-edit
    ",s" 'slack-message-send-from-buffer
    ",2" 'slack-message-embed-mention
    ",3" 'slack-message-embed-channel))

(defun prefer-horizontal-split ()
  (set-variable 'split-height-threshold nil t)
  (set-variable 'split-width-threshold 40 t)) ; make this as low as needed
(add-hook 'markdown-mode-hook 'prefer-horizontal-split)
(map! :leader
      :desc "Clone indirect buffer other window"
      "b c" #'clone-indirect-buffer-other-window)

(setq telega-use-docker t
      telega-use-images t)

(map! :leader
      :desc "Winner redo"
      "w <right>" #'winner-redo
      :leader
      :desc "Winner undo"
      "w <left>" #'winner-undo)
